### ISSUE TITLE
Timing Delay

### ISSUE THREAD

**Issue by FabioCoder** [astronomer/airflow-provider-fivetran-async#32]
Hi Folks,

We recently started using the FivetranOperatorAsync operator because the FivetranOperator is deprecated and cannot handle reschedules.

We are observing large delays between the task initialization and the execution. Here is an excerpt from the logs:

### Run ID: scheduled__2023-06-27T17:00:00+00:00

```
...
[2023-06-27, 20:00:03 CEST] {taskinstance.py:1415} INFO - Pausing task as DEFERRED. dag_id=transactions_speed_data_to_snowflake_dag, task_id=sync_mysql_c24_main_speed, execution_date=20230627T170000, start_date=20230627T180002
[2023-06-27, 20:00:03 CEST] {local_task_job_runner.py:222} INFO - Task exited with return code 100 (task deferral)
[2023-06-27, 22:17:21 CEST] {base.py:73} INFO - Using connection ID 'fivetran_chronolytics' for task execution.
[2023-06-27, 22:17:23 CEST] {hooks.py:129} INFO - Connector "plied_presenting": sync_state = "scheduled"
[2023-06-27, 22:17:23 CEST] {triggers.py:89} INFO - sync is still running...
[2023-06-27, 22:17:23 CEST] {triggers.py:90} INFO - sleeping for 15 seconds.
[2023-06-27, 22:17:41 CEST] {hooks.py:129} INFO - Connector "plied_presenting": sync_state = "scheduled"
...

[2023-06-27, 22:46:57 CEST] {triggerer_job_runner.py:608} INFO - Trigger transactions_speed_data_to_snowflake_dag/scheduled__2023-06-27T17:00:00+00:00/sync_mysql_c24_main_speed/-1/1 (ID 452) fired: TriggerEvent<{'status': 'error', 'message': "Cannot connect to host api.fivetran.com:443 ssl:default [Connect call failed ('35.236.237.87', 443)]"}>
[2023-06-27, 22:47:00 CEST] {taskinstance.py:1103} INFO - Dependencies all met for dep_context=non-requeueable deps ti=<TaskInstance: transactions_speed_data_to_snowflake_dag.sync_mysql_c24_main_speed scheduled__2023-06-27T17:00:00+00:00 [queued]>
[2023-06-27, 22:47:00 CEST] {taskinstance.py:1103} INFO - Dependencies all met for dep_context=requeueable deps ti=<TaskInstance: transactions_speed_data_to_snowflake_dag.sync_mysql_c24_main_speed scheduled__2023-06-27T17:00:00+00:00 [queued]>
[2023-06-27, 22:47:00 CEST] {taskinstance.py:1306} INFO - Resuming after deferral
[2023-06-27, 22:47:00 CEST] {taskinstance.py:1327} INFO - Executing <Task(FivetranOperatorAsync): sync_mysql_c24_main_speed> on 2023-06-27 17:00:00+00:00
[2023-06-27, 22:47:00 CEST] {standard_task_runner.py:57} INFO - Started process 1655824 to run task
[2023-06-27, 22:47:00 CEST] {standard_task_runner.py:84} INFO - Running: ['airflow', 'tasks', 'run', 'transactions_speed_data_to_snowflake_dag', 'sync_mysql_c24_main_speed', 'scheduled__2023-06-27T17:00:00+00:00', '--job-id', '331794', '--raw', '--subdir', 'DAGS_FOLDER/extract/fivetran_data_to_snowflake_dag.py', '--cfg-path', '/tmp/tmpdhc5q_08']
[2023-06-27, 22:47:00 CEST] {standard_task_runner.py:85} INFO - Job 331794: Subtask sync_mysql_c24_main_speed
[2023-06-27, 22:47:00 CEST] {task_command.py:410} INFO - Running <TaskInstance: transactions_speed_data_to_snowflake_dag.sync_mysql_c24_main_speed scheduled__2023-06-27T17:00:00+00:00 [running]> on host airflow
[2023-06-27, 22:47:00 CEST] {taskinstance.py:1824} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.9/site-packages/fivetran_provider_async/operators.py", line 92, in execute_complete
    raise AirflowException(msg)
airflow.exceptions.AirflowException: error: Cannot connect to host api.fivetran.com:443 ssl:default [Connect call failed ('35.236.237.87', 443)]
...
```

### **Run ID: scheduled__2023-06-27T18:00:00+00:00**

```
...
[2023-06-27, 21:00:05 CEST] {taskinstance.py:1415} INFO - Pausing task as DEFERRED. dag_id=transactions_speed_data_to_snowflake_dag, task_id=sync_mysql_c24_main_speed, execution_date=20230627T180000, start_date=20230627T190002
[2023-06-27, 21:00:05 CEST] {local_task_job_runner.py:222} INFO - Task exited with return code 100 (task deferral)
[2023-06-27, 22:17:21 CEST] {base.py:73} INFO - Using connection ID 'fivetran_chronolytics' for task execution.
[2023-06-27, 22:17:23 CEST] {hooks.py:129} INFO - Connector "plied_presenting": sync_state = "scheduled"
[2023-06-27, 22:17:23 CEST] {triggers.py:89} INFO - sync is still running...
[2023-06-27, 22:17:23 CEST] {triggers.py:90} INFO - sleeping for 15 seconds.
[2023-06-27, 22:17:41 CEST] {hooks.py:129} INFO - Connector "plied_presenting": sync_state = "scheduled"
...

[2023-06-27, 22:47:02 CEST] {triggerer_job_runner.py:608} INFO - Trigger transactions_speed_data_to_snowflake_dag/scheduled__2023-06-27T18:00:00+00:00/sync_mysql_c24_main_speed/-1/1 (ID 455) fired: TriggerEvent<{'status': 'error', 'message': "Cannot connect to host api.fivetran.com:443 ssl:default [Connect call failed ('35.236.237.87', 443)]"}>
[2023-06-27, 22:47:04 CEST] {taskinstance.py:1103} INFO - Dependencies all met for dep_context=non-requeueable deps ti=<TaskInstance: transactions_speed_data_to_snowflake_dag.sync_mysql_c24_main_speed scheduled__2023-06-27T18:00:00+00:00 [queued]>
[2023-06-27, 22:47:04 CEST] {taskinstance.py:1103} INFO - Dependencies all met for dep_context=requeueable deps ti=<TaskInstance: transactions_speed_data_to_snowflake_dag.sync_mysql_c24_main_speed scheduled__2023-06-27T18:00:00+00:00 [queued]>
[2023-06-27, 22:47:04 CEST] {taskinstance.py:1306} INFO - Resuming after deferral
[2023-06-27, 22:47:04 CEST] {taskinstance.py:1327} INFO - Executing <Task(FivetranOperatorAsync): sync_mysql_c24_main_speed> on 2023-06-27 18:00:00+00:00
[2023-06-27, 22:47:04 CEST] {standard_task_runner.py:57} INFO - Started process 1655892 to run task
[2023-06-27, 22:47:04 CEST] {standard_task_runner.py:84} INFO - Running: ['airflow', 'tasks', 'run', 'transactions_speed_data_to_snowflake_dag', 'sync_mysql_c24_main_speed', 'scheduled__2023-06-27T18:00:00+00:00', '--job-id', '331796', '--raw', '--subdir', 'DAGS_FOLDER/extract/fivetran_data_to_snowflake_dag.py', '--cfg-path', '/tmp/tmp7y56_u6y']
[2023-06-27, 22:47:04 CEST] {standard_task_runner.py:85} INFO - Job 331796: Subtask sync_mysql_c24_main_speed
[2023-06-27, 22:47:04 CEST] {task_command.py:410} INFO - Running <TaskInstance: transactions_speed_data_to_snowflake_dag.sync_mysql_c24_main_speed scheduled__2023-06-27T18:00:00+00:00 [running]> on host airflow
[2023-06-27, 22:47:04 CEST] {taskinstance.py:1824} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.9/site-packages/fivetran_provider_async/operators.py", line 92, in execute_complete
    raise AirflowException(msg)
airflow.exceptions.AirflowException: error: Cannot connect to host api.fivetran.com:443 ssl:default [Connect call failed ('35.236.237.87', 443)]
...
```

### **Run ID: scheduled__2023-06-27T19:00:00+00:00**

```
...
[2023-06-27, 22:00:04 CEST] {taskinstance.py:1415} INFO - Pausing task as DEFERRED. dag_id=transactions_speed_data_to_snowflake_dag, task_id=sync_mysql_c24_main_speed, execution_date=20230627T190000, start_date=20230627T200002
[2023-06-27, 22:00:04 CEST] {local_task_job_runner.py:222} INFO - Task exited with return code 100 (task deferral)
[2023-06-27, 22:17:21 CEST] {base.py:73} INFO - Using connection ID 'fivetran_chronolytics' for task execution.
[2023-06-27, 22:17:23 CEST] {hooks.py:129} INFO - Connector "plied_presenting": sync_state = "scheduled"
[2023-06-27, 22:17:23 CEST] {triggers.py:89} INFO - sync is still running...
[2023-06-27, 22:17:23 CEST] {triggers.py:90} INFO - sleeping for 15 seconds.
[2023-06-27, 22:17:41 CEST] {hooks.py:129} INFO - Connector "plied_presenting": sync_state = "scheduled"
...

[2023-06-27, 22:47:35 CEST] {triggerer_job_runner.py:608} INFO - Trigger transactions_speed_data_to_snowflake_dag/scheduled__2023-06-27T19:00:00+00:00/sync_mysql_c24_main_speed/-1/1 (ID 458) fired: TriggerEvent<{'status': 'error', 'message': "Cannot connect to host api.fivetran.com:443 ssl:default [Connect call failed ('35.236.237.87', 443)]"}>
[2023-06-27, 22:47:36 CEST] {taskinstance.py:1103} INFO - Dependencies all met for dep_context=non-requeueable deps ti=<TaskInstance: transactions_speed_data_to_snowflake_dag.sync_mysql_c24_main_speed scheduled__2023-06-27T19:00:00+00:00 [queued]>
[2023-06-27, 22:47:36 CEST] {taskinstance.py:1103} INFO - Dependencies all met for dep_context=requeueable deps ti=<TaskInstance: transactions_speed_data_to_snowflake_dag.sync_mysql_c24_main_speed scheduled__2023-06-27T19:00:00+00:00 [queued]>
[2023-06-27, 22:47:36 CEST] {taskinstance.py:1306} INFO - Resuming after deferral
[2023-06-27, 22:47:36 CEST] {taskinstance.py:1327} INFO - Executing <Task(FivetranOperatorAsync): sync_mysql_c24_main_speed> on 2023-06-27 19:00:00+00:00
[2023-06-27, 22:47:36 CEST] {standard_task_runner.py:57} INFO - Started process 1656304 to run task
[2023-06-27, 22:47:36 CEST] {standard_task_runner.py:84} INFO - Running: ['airflow', 'tasks', 'run', 'transactions_speed_data_to_snowflake_dag', 'sync_mysql_c24_main_speed', 'scheduled__2023-06-27T19:00:00+00:00', '--job-id', '331798', '--raw', '--subdir', 'DAGS_FOLDER/extract/fivetran_data_to_snowflake_dag.py', '--cfg-path', '/tmp/tmpe4dkt18r']
[2023-06-27, 22:47:36 CEST] {standard_task_runner.py:85} INFO - Job 331798: Subtask sync_mysql_c24_main_speed
[2023-06-27, 22:47:36 CEST] {task_command.py:410} INFO - Running <TaskInstance: transactions_speed_data_to_snowflake_dag.sync_mysql_c24_main_speed scheduled__2023-06-27T19:00:00+00:00 [running]> on host airflow
[2023-06-27, 22:47:36 CEST] {taskinstance.py:1824} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.9/site-packages/fivetran_provider_async/operators.py", line 92, in execute_complete
    raise AirflowException(msg)
airflow.exceptions.AirflowException: error: Cannot connect to host api.fivetran.com:443 ssl:default [Connect call failed ('35.236.237.87', 443)]
...

```

I have built a small working example from the code we use. Since we need the operator several times, I have outsourced the initialisation to a method.

```python

import pendulum
from airflow import DAG
from fivetran_provider_async.operators import FivetranOperatorAsync

FIVETRAN_CONN_ID = 'xxx'
FIVETRAN_DAG_NAME_TEMPLATE = "{}_data_to_snowflake_dag"
FIVETRAN_DAG_ARGS = {
    "start_date": pendulum.today('UTC').add(days=-1),
    "catchup": False
}
CONNECTORS = [{'name':'xxx', 'id': 'xxx'}]

def sync_connectors(connectors, dag=None):

    for connector in connectors:

        run_fivetran_sync = FivetranOperatorAsync(
            task_id=f"sync_{connector['name'].lower()}",
            fivetran_conn_id=FIVETRAN_CONN_ID,
            connector_id=connector['id'],
            dag=dag
        )

    return run_fivetran_sync

sync_transactions_hourly_data_dag = DAG(
    dag_id=FIVETRAN_DAG_NAME_TEMPLATE.format('transactions_speed'),
    schedule="0 * * * *",
    **FIVETRAN_DAG_ARGS
)

sync_connectors(connectors, dag=sync_transactions_hourly_data_dag)

```

We host Airflow ourselves and have not used Deferred Tasks so far. Therefore, I cannot exclude that there is a problem with the configuration of the environment. Can anyone help here? 

Best,
Fabio

**Comment by FabioCoder** [comment]
A similar problem is probably described in #24.

**Comment by sunank200** [comment]
Could you share the steps to replicate this issue please?

**Comment by sunank200** [comment]
Also what is the connection type for this?

**Comment by FabioCoder** [comment]
I cannot provide exact steps to recreate the problem. 

The problem occurred when about 15 tasks were executed with this operator defined in different dags.

We have temporarily switched back to the Fivetran operator in combination with the Fivetran sensor for the majority of our tasks. Since then, the problem has no longer occurred.  So I assume that the number of tasks running in parallel has an influence on the issue.

**Comment by FabioCoder** [comment]
The Connection type is fivetran.

**Comment by sunank200** [comment]
I am running the multiple DAGs as mentioned in the example in our astro cloud: https://astro01.astronomer-stage.run/dko7l7m7/home

But the DAG seems to run fine without any errors. I am trying to replicate this.

**Comment by sunank200** [comment]
Meanwhile, as this [PR](https://github.com/astronomer/airflow-provider-fivetran-async/pull/33) is merged `FivetranOperator` can be used as part of the next release.

**Comment by phanikumv** [comment]
@FabioCoder could you kindly confirm if this issue is resolved for you?

**Comment by vatsrahul1001** [comment]
@FabioCoder  can you kindly confirm if issue persists?

**Comment by FabioCoder** [comment]
Hi, I can't confirm that the problem is solved. Because of the big problems we had, we switched back to the package provided by fivetran. To ensure that the rescheduling is done by the Fivetran engine, we set the schedule type to 'auto' after starting a sync. This setup works fine.

**Comment by FabioCoder** [comment]
I will close the ticket anyway. Because we have found a workaround.
